strutil/set.go

```diff
@@ -0,0 +1,96 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+// OrderedSet is a set of strings that maintains the order of insertion.
+//
+// External synchronization is required for safe concurrent access.
+type OrderedSet struct {
+	positionOf map[string]int
+}
+
+// Items returns a slice of strings representing insertion order.
+//
+// Contains is O(N) in the size of the set.
+func (o *OrderedSet) Items() []string {
+	if o.positionOf == nil {
+		return nil
+	}
+	items := make([]string, len(o.positionOf))
+	for item, idx := range o.positionOf {
+		items[idx] = item
+	}
+	return items
+}
+
+// Contains returns true if the set contains a given item.
+//
+// Contains is O(1) in the size of the set.
+func (o *OrderedSet) Contains(item string) bool {
+	if o.positionOf == nil {
+		return false
+	}
+
+	_, ok := o.positionOf[item]
+	return ok
+}
+
+// IndexOf returns the position of an item in the set.
+func (o *OrderedSet) IndexOf(item string) (idx int, ok bool) {
+	idx, ok = o.positionOf[item]
+	return idx, ok
+}
+
+// Del removes an item from the set.
+//
+// Delete is O(1) in the size of the set.
```

Comment:
del, or delete?

Reply:
Yeah, I prefer Del because it's shorter and matches the other data structure we have but Delete is more like the go built-in. I'll unify to Del but we can swap to Delete later.

Reply:
Del/Delete is gone now